module/zfs/ddt.c

```diff
@@ -575,9 +596,51 @@ ddt_remove(ddt_t *ddt, ddt_entry_t *dde)
 	ddt_free(dde);
 }
 
+/*
+ * Check if the DDT is over its quota.  This can be due to a few conditions:
+ *   1. 'dedup_table_quota' is 0 and ddt_get_ddt_dsize exceeds this limit
+ *
+ *   2. 'dedup_table_quota' is set to automatic and
+ *      a. the dedup allocation class could not satisfy a DDT allocation
+ *         in a recent transaction
+ *      b. the dedup allocation class has exceeded its calculated limit
+ */
+static boolean_t
+ddt_over_quota(spa_t *spa)
+{
+	if (spa->spa_dedup_table_quota == 0)
+		return (B_FALSE);
+
+	if (spa->spa_dedup_table_quota != UINT64_MAX)
+		return (ddt_get_ddt_dsize(spa) > spa->spa_dedup_table_quota);
+
+	/*
+	 * For automatic quota, table size is limited by the size of dedup class
+	 */
+	metaslab_class_t *dedup_mc = spa_dedup_class(spa);
+	if (dedup_mc != NULL && metaslab_class_get_space(dedup_mc) > 0) {
+		/* Over quota if allocating outside of dedup allocation class */
+		if (spa_syncing_txg(spa) <= spa->spa_dedup_class_full_txg +
+		    dedup_class_wait_txgs) {
+			/* Waiting for deferred frees to be processed */
+			return (B_TRUE);
+		}
+
+		/* Use larger of 85% of space or space - 8G */
+		uint64_t allocated = metaslab_class_get_alloc(dedup_mc);
+		uint64_t capacity = metaslab_class_get_space(dedup_mc);
+		uint64_t limit = MAX(capacity * 85 / 100, capacity - (1LL<<33));
```

Comment:
It seems formally ZFS allows vdevs of 64MB.  In that case `capacity - (1LL<<33)` will give negative overflow.  I also wonder why 8GB?  Is this such a way to ignore 85% requirement for devices below 8GB?  In the test below I see 200MB device used.

Reply:
Good catch.  the code/comments were unclear.

We're considered over quota when we hit 85% full, or for larger drives, when there is less than 8GB free. Updated code  & comments